wagtail/admin/rich_text/converters/html_to_contentstate.py

```diff
@@ -0,0 +1,288 @@
+            state.current_block.text += ' '
+            state.leading_whitespace = 'strip'
+
+        inline_style_range = InlineStyleRange(self.style)
+        inline_style_range.offset = len(state.current_block.text)
+        state.current_block.inline_style_ranges.append(inline_style_range)
+        state.current_inline_styles.append(inline_style_range)
+
+    def handle_endtag(self, name, state, contentstate):
+        inline_style_range = state.current_inline_styles.pop()
+        assert inline_style_range.style == self.style
+        inline_style_range.length = len(state.current_block.text) - inline_style_range.offset
+
+
+class LinkElementHandler(object):
+    def __init__(self, entity_type):
+        self.entity_type = entity_type
+
+    def handle_starttag(self, name, attrs, state, contentstate):
+        assert state.current_block is not None, "%s element found at the top level" % name
+
+        if state.leading_whitespace == 'force':
+            # any pending whitespace should be output before handling this tag,
+            # and subsequent whitespace should be collapsed into it (= stripped)
+            state.current_block.text += ' '
+            state.leading_whitespace = 'strip'
+
+        attrs = dict(attrs)
+
+        entity = Entity(self.entity_type, 'MUTABLE', self.get_attribute_data(attrs))
+        key = contentstate.add_entity(entity)
+
+        entity_range = EntityRange(key)
+        entity_range.offset = len(state.current_block.text)
+        state.current_block.entity_ranges.append(entity_range)
+        state.current_entity_ranges.append(entity_range)
+
+    def get_attribute_data(self, attrs):
+        return {}
+
+    def handle_endtag(self, name, state, contentstate):
+        entity_range = state.current_entity_ranges.pop()
+        entity_range.length = len(state.current_block.text) - entity_range.offset
+
+
+class ExternalLinkElementHandler(LinkElementHandler):
+    def get_attribute_data(self, attrs):
+        return {'url': attrs['href']}
+
+
+class PageLinkElementHandler(LinkElementHandler):
+    def get_attribute_data(self, attrs):
+        try:
+            page = Page.objects.get(id=attrs['id']).specific
+        except Page.DoesNotExist:
+            return {}
+
+        data = {
+            'id': page.id,
+            'url': page.url,
+        }
+
+        parent_page = page.get_parent()
+        if parent_page:
+            data['parentId'] = parent_page.id
+
+        return data
+
+
+class AtomicBlockEntityElementHandler(object):
+    """
+    Handler for elements like <img> that exist as a single immutable item at the block level
+    """
+    def handle_starttag(self, name, attrs, state, contentstate):
+        # forcibly close any block that illegally contains this one
+        state.current_block = None
+
+        entity = self.create_entity(name, dict(attrs), state, contentstate)
+        key = contentstate.add_entity(entity)
+
+        block = Block('atomic', depth=state.list_depth)
```

Comment:
Draft.js here.

Reply:
I see. I'm not sure what's the best thing to do then. Hardcodding it with `0` for now (and adding a comment) might mess with the whole state on the Python side (getting `push/pop` actions and depth out of sync?) but I'm not sure how the front-end might react if we leave it like this for now?

Reply:
The Python code doesn't care what we put here - we track `list_depth` as part of HandlerState, but once we create a Block and assign it a `depth` property we never read that back, so we could equally pass `0` here with no ill effects.

The scenario where this comes up would be something like:

    <ul>
        <li>level 1 item 1</li>
        <li>level 1 item 2
            <ul>
                <li>level 2 item 1</li>
                <li><embed embedtype="image" id="1" format="left" /></li>
                <li>level 2 item 3</li>
            </ul>
        </li>
        <li>level 1 item 3</li>
    </ul>

I'd consider this to be very much an edge case, not least because hallo.js can't do multi-level lists itself - this would only arise by pasting HTML from outside and then editing the results, if it's possible at all. We don't expect Draftail to be able to preserve this structure either, so I guess it comes down to the question: which output is Draftail less likely to trip up on, an atomic block with depth>0 or a block with depth=0 in the middle of a list?